design/resource-dependency.md

```diff
@@ -0,0 +1,348 @@
+      and passed to EtcdBackup CR field `spec.connSecret`.
+
+### Limitations
+
+1. All the resources in the dependency graph should be created and managed by OAM runtime.
+2. This proposal is designed only for creating resources. After resources have been created, subsequent data output updates will not trigger any actions.
+
+### Generic core engine consideration
+
+The design of resource dependency is a pretty generic engine that works for any k8s resources.
+Even though we use OAM native fields like `Parameters` or `Outputs` fields, those are just representation layer.
+The underlying dependency engine along with the DataInput and DataOutput model is the core implementation,
+which should be generic enough to convert or adopt future concepts or fields.
+
+### Spec upgrade path consideration
+
+Initially the feature will be experimental. It might change iteratively.
+Once that happens, we strive to keep it upgradeable between two releases.
+Here we discuss an idea to upgrade if that happens.
+
+The core dependency engine in OAM runtime is a generic implementation.
+We would map APIs to underlying DataInput/DataOutput objects.
+
+It's inevitable that an intermediate phase where both old and new APIs co-exist.
+But they don't both exist in the same ApplicationConfiguration.
+Only one could exist and we don't need to handle any conflicts.
+In that case, We could add an webhook to translate the old to the new API.
+
+## Solution examples
+
+In this section we are providing more solution examples for dependency use cases.
+
+1. Web app and database.
+
+    ```yaml
+    kind: Component
+    metadata:
+      name: my-rds
+    spec:
+      workload:
+        kind: RDSInstance
+        metadata:
+          name: my-rds
+    ---
+
+    kind: Component
+    metadata:
+      name: my-web
+    spec:
+      workload:
+        kind: MyWorkload
+        metadata:
+          name: my-web
+      parameters:
+      - name: database-secret
+        fieldPaths: ["spec.connSecret"]
+    ---
+
+    kind: ApplicationConfiguration
+    spec:
+      components:
+      - componentName: my-web
+        parameterValues:
+        - name: database-secret
+          valueFrom:
+            dataOutputName: mysql-conn-secret
+      - componentName: my-rds
+        dataOutputs:
+          - name: mysql-conn-secret
+            fieldPath: "status.connSecret"
+    ```
+
+2. Service binding trait.
+
+    ApplicationConfiguration :
+
+    ```yaml
+    kind: ApplicationConfiguration
+    spec:
+      components:
+      - componentName: my-web
+        traits:
+        - trait:
+            apiVersion: core.oam.dev/v1alpha1
+            kind: ServiceBinding
+            metadata:
+              name: my-secret-binding
+            spec:
+              from:
+                secret:
+                  name: my-secret
+              to:
+                env: true
+          dataOutputs:
+            - name: my-secret-binding
+              fieldPath: status.ready
+    ```
```

Comment:
why is there no `parameterValues` field that contains `valueFrom`  in the component here? Is it just implied?

Reply:
I get that this is pretty much a `syntax sugar` which basically avoids writing another `parameterValues` in the component. I am not sure if I like it or not. On one hand, it saves some repetitive work but on the other hand, it is inconsistent with the way parameters are assigned in components.